src/components/calendar.rs

```diff
@@ -1,28 +1,263 @@
+        div class="flex-1 flex flex-col" {
+            div class="text-[calc(1vh)] lg:text-sm shrink flex justify-center items-center" { (name) }
+            div class="relative flex flex-col grow gap-0.5 lg:gap-1" {
+                @for _ in timeslots {
+                    div class="h-auto grow bg-neutral-600" {  }
+                }
+                @for section in sections {
+                    (render_section_cards(earliest, latest, &section))
+                }
+            }
+        }
+    )
+}
+
+fn time_to_string(time: f64) -> String {
+    let absolute_hour = time.floor();
+    let hour = if absolute_hour >= 13.0 {
+        absolute_hour - 12.0
+    } else {
+        absolute_hour
+    };
+    let minute = if time - absolute_hour == 0.5 {
+        "30"
+    } else {
+        "00"
+    };
+    let meridiem = if time >= 12.0 { "pm" } else { "am" };
+    format!("{hour}:{minute}{meridiem}")
+}
+
+pub fn render(sections: &Vec<Section>) -> Markup {
+    debug!(?sections);
+
+    let tmp_meeting = Section {
+        crn: 12015,
+        subject_code: "MATH".to_string(),
+        course_code: "100".to_string(),
+        sequence_code: "A01".to_string(),
+        enrollment: 0,
+        enrollment_capacity: 30,
+        waitlist: 0,
+        waitlist_capacity: 10,
+        meeting_times: vec![MeetingTime {
+            start_date: "2024-01-01".parse().unwrap(),
+            start_time: Some("08:30:00".parse().unwrap()),
+            end_date: "2024-04-30".parse().unwrap(),
+            end_time: Some("11:20:00".parse().unwrap()),
+            days: Days {
+                monday: true,
+                tuesday: false,
+                wednesday: false,
+                thursday: true,
+                friday: false,
+                saturday: false,
+                sunday: false,
+            },
+            building: Some("ECS".to_string()),
+            room: Some("123".to_string()),
+        }],
+    };
+
+    let tmp_meeting2 = Section {
+        crn: 12072,
+        subject_code: "CSC".to_string(),
+        course_code: "111".to_string(),
+        sequence_code: "A01".to_string(),
+        enrollment: 30,
+        enrollment_capacity: 30,
+        waitlist: 0,
+        waitlist_capacity: 10,
+        meeting_times: vec![MeetingTime {
+            start_date: "2024-01-01".parse().unwrap(),
+            start_time: Some("16:30:00".parse().unwrap()),
+            end_date: "2024-04-30".parse().unwrap(),
+            end_time: Some("17:20:00".parse().unwrap()),
+            days: Days {
+                monday: false,
+                tuesday: true,
+                wednesday: true,
+                thursday: false,
+                friday: true,
+                saturday: false,
+                sunday: false,
+            },
+            building: Some("ECS".to_string()),
+            room: Some("123".to_string()),
+        }],
+    };
+    let sections = vec![tmp_meeting, tmp_meeting2];
+
+    let earliest = match sections
```

Comment:
Not a huge fan of using numeric types to represent time. We have the `jiff` datetime library so we might as well use it!

How about something like this?

```rust
let earliest = sections
    .iter()
    .flat_map(|s| &s.meeting_times)
    .flat_map(|t| t.start_time)
    .min()
    .unwrap_or(jiff::civil::time(8, 0, 0, 0));
```
Then we can create the timeslots like this:
```rust
let timeslots = {
    let first_slot = earliest
        .round(
            TimeRound::new()
                .smallest(jiff::Unit::Minute)
                .increment(30)
                .mode(jiff::RoundMode::Floor),
        )
        .unwrap();

    let last_slot = latest
        .round(
            TimeRound::new()
                .smallest(jiff::Unit::Minute)
                .increment(30)
                .mode(jiff::RoundMode::Ceil),
        )
        .unwrap();

    first_slot
        .series(30.minutes())
        .take_while(|&t| t <= last_slot)
        .collect::<Vec<_>>()
};
```

Other stuff like using `strftime` for formatting and the built-in math and rounding could make the current code shorter.

Reply:
On second thought, given how we're using `earliest` and `latest` for formatting and such, it might be better to move the rounding to those instead of in `timeslots`'s declaration so it doesn't break stuff 😬. My bad for the recommendation.